pulsar-function-go/pf/instance.go

```diff
@@ -0,0 +1,289 @@
+					SubscriptionName:  subscriptionName,
+					Properties:        properties,
+					Type:              consumerType,
+					MessageChannel:    channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+				gi.consumers = append(gi.consumers, consumer)
+			} else {
+				consumer, err := gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:             topic,
+					SubscriptionName:  subscriptionName,
+					Properties:        properties,
+					Type:              consumerType,
+					ReceiverQueueSize: int(consumerConf.ReceiverQueueSize.Value),
+					MessageChannel:    channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+
+				gi.consumers = append(gi.consumers, consumer)
+			}
+		} else {
+			if consumerConf.IsRegexPattern {
+				consumer, err := gi.client.Subscribe(pulsar.ConsumerOptions{
+					TopicsPattern:    topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             consumerType,
+					MessageChannel:   channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+				gi.consumers = append(gi.consumers, consumer)
+			} else {
+				consumer, err := gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:            topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             consumerType,
+					MessageChannel:   channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+				gi.consumers = append(gi.consumers, consumer)
+			}
+		}
+		gi.context.InputTopics = append(gi.context.InputTopics, topic)
+
+	CLOSE:
+		for {
+			select {
+			case cm := <-channel:
+				gi.msg = append(gi.msg, cm.Message)
+			case <-time.After(time.Millisecond * 500):
+				close(channel)
+				break CLOSE
+			}
+		}
+	}
+
+	gi.actualExecution(function)
+	gi.close()
+}
+
+func (gi *GoInstance) actualExecution(function Function) {
+	gi.function = function
+	output, err := gi.handlerMsg()
+	err = gi.processResult(output)
+	if err != nil {
+		log.Errorf("process output result err:%v", err)
+		return
+	}
+}
+
+func (gi *GoInstance) processResult(output []byte) error {
+	gi.context = NewFuncContext()
+	atLeastOnce := gi.context.InstanceConf.FuncDetails.ProcessingGuarantees == pb.ProcessingGuarantees_ATLEAST_ONCE
+
+	if output != nil && gi.context.InstanceConf.FuncDetails.Sink.Topic != "" &&
+		len(gi.context.InstanceConf.FuncDetails.Sink.Topic) > 0 {
+		if gi.producer == nil {
+			err := gi.setUpProducer()
+			if err != nil {
+				log.Errorf("set up producer error:%s", err.Error())
+				return err
+			}
+		}
+
+		asyncMsg := pulsar.ProducerMessage{
+			Payload: output,
+		}
+		// Attempt to send the message asynchronously and handle the response
+		gi.producer.SendAsync(context.Background(), asyncMsg, func(message pulsar.ProducerMessage, e error) {
+			if e != nil {
+				log.Fatal(e)
```

Comment:
if failed to write the message and processing guarantee is >= at-least-once, you can negative-ack the message.

Reply:
actually the logic here should be
if (autoack) {
    1. if atleast-once;- on error nack, else ack
    2. if effectively-once:- on error fatal

Reply:
I still have doubts about this logic.
If there is output;-
    write, on success if autoAck ack
               on failure nack(or die if effectively once)
else:- if autoack && !atmostOnce:- ack

Reply:
@srkukarni thanks.

```
if (output != nil){
    if autoAck {
	 if e != nil {
		 if atLeastOnce {
			 gi.nackInputMessage(msgInput)
		 } else if effectivelyOnce {
			 panic("in effectively-once case, this is error.")
            }
    } else if !atMostOnce {
		gi.ackInputMessage(msgInput)
	    }
     }
} else {
    if autoAck {
	gi.ackInputMessage(msgInput)
    }
}
```